### ISSUE TITLE
Misinterpretation of 32-bit addresses for 16-bit application

### ISSUE THREAD

**Issue by Wall-AF** [NationalSecurityAgency/ghidra#2548]
**Describe the bug**
When looking at a 16-bit application Ghidra will recognise 32-bit pointers, however it is swapping the segment and offset incorrectly in `PointerDataType::getSegmentedAddressValue(...)`.

**To Reproduce**
1. Open a Windows NE format application/DLL.
2. Find a 32-bit pointer reference.
3. Try to turn it into a 32-bit pointer (by any means).
4. See the address in to label (and internally) is incorrect.
5. Try following that label and it'll either fail or goto the incorrect location.

**Expected behavior**
Get the correct address.

**Example (Screenshots)**
```
                             FnLookupTable_1210_0e36
       1210:0e36 fe 10 10 12     FARPROC    DAT_10fe_1210
       1210:0e3a 03 11 10 12     FARPROC    DAT_1103_1210
       1210:0e3e 08 11 10 12     FARPROC    LAB_1108_1210
```
instead of:
```
                             FnLookupTable_1210_0e36
       1210:0e36 fe 10 10 12     FARPROC    DAT_1210_10fe
       1210:0e3a 03 11 10 12     FARPROC    DAT_1210_1103
       1210:0e3e 08 11 10 12     FARPROC    LAB_1210_1108
```

**Environment (please complete the following information):**
 - OS: Windows 10 (20H2)
 - Java Version: e.g. 11.0.8
 - Ghidra Version: 9.2

**Fix**
Modify this code snippet from `PointerDataType::getSegmentedAddressValue(...)`:
```
				case 4: // far pointer
					segment = buf.getUnsignedShort(0);
					offset = buf.getUnsignedShort(2);
					break;
```
into
```
				case 4: // far pointer
					offset = buf.getUnsignedShort(0);
					segment = buf.getUnsignedShort(2);
					break;
```

**Comment by Wall-AF** [comment]
Here is a more robust replacement:
```
					if (buf.isBigEndian()) {
						segment = buf.getUnsignedShort(0);
						offset = buf.getUnsignedShort(2);
					} else {
						offset = buf.getUnsignedShort(0);
						segment = buf.getUnsignedShort(2);
					}
```

**Comment by ghidra1** [comment]
> Here is a more robust replacement:
> 
> ```
> 					if (buf.isBigEndian()) {
> 						segment = buf.getUnsignedShort(0);
> 						offset = buf.getUnsignedShort(2);
> 					} else {
> 						offset = buf.getUnsignedShort(0);
> 						segment = buf.getUnsignedShort(2);
> 					}
> ```

The MemBuffer already accounts for endianess.  In addition, all of the segmented implementations only target little-endian x86.

**Comment by Wall-AF** [comment]
The original  code takes 4 bytes, 2 for the segment and 2 for the offset, but crucially the wrong  way round.

**Comment by ghidra1** [comment]
Based upon the slaspec (ia.sinc) code below it certainly would seem segment comes after offset:
```
ptr1616: reloc is protectedMode=0 & imm16; j16		[ reloc = j16*0x10 + imm16; ] { CS = j16; export *[ram]:4 reloc; }
ptr1616: reloc is protectedMode=1 & imm16; j16		[ reloc = j16*0x10000 + imm16; ] { CS = j16; export *[ram]:4 reloc; }
```
While the spec indicates the following, I believe this reflects the little-endian byte swap on the 32-bit stored value.  Whereas the pcode is reading 16-bits at a time.  Offset first followed by segment.
```
CS ← DEST[31:16]; (* DEST is ptr16:16 or [m16:16] *)
EIP ← DEST[15:0]; (* DEST is ptr16:16 or [m16:16]; clear upper 16 bits *)
```

**Comment by Wall-AF** [comment]
If the address was read using `addr = buf.getUnsignedLong(0)` and split, it would be correct, but the original code isn't.  Also, 32-bit addresses are held low to high order.

**Comment by ghidra1** [comment]
@Wall-AF , I get it.  The pointer implementation is certainly inconsistent with Ghidra's language implementation I showed above (`ptr1616`).   I don't think we need the endianess logic you provided above.  The order of the two short reads just need to be reversed as you first indicated.

**Comment by Wall-AF** [comment]
Okay, just wanted to be thorough in case of reuse elsewhere.

Thanks.

**Comment by ghidra1** [comment]
Corrected for 9.2.3
